### PR TITLE
PR: Close leaks without changing tests

### LINKED ISSUE #12328: test_mainwindow.py opens too many file descriptors and sockets
Hi,

First, I want to clarify that I am fully aware of the fact, that you do not support running unit tests outside of your own development environments and CIs. However, because I think this might be related to what happens in #12247, I am reporting anyway. Please, don't feel obliged to dig too much into it if you feel this does not apply to the scenarios you are officially supporting.

In my endeavor to stabilize Spyder for the OpenSUSE Linux distro package which I am maintaining, I try to run as many unit tests as possible in order to detect possible package incompatibilites.

Fact is, the execution of `python3 runtests.py --run-slow` always fails with a segfault or breakpoint trap. This happens not always at the same test but consistently a few dozen tests into the test_mainwindow.py suite. *Sometimes* there is a `Too many open files` error message associated with it.

I added a few debug statements to the `main_window` fixture and found out that for each test, the number of open files and sockets increases. Specifically the `pdb_history.sqlite` file is opened many times but not closed at teardown of the test. I suspect this is because of the reuse of the window as introduced in #10444. This could also be the reason why #12247 cannot open any new connections.

I will dig more into this, but any hints would be appreciated.

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

This is #12340  without the tests change

### Issue(s) Resolved

Fixes #12328

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by impact27** [comment]
@bnavigator could you try this PR?

**Comment by bnavigator** [comment]
Could you please change the target branch from `master` to `4.x`? Your local branch is based on that.

I had to rebase to `v4.1.2` in order to test it with the package that I am trying to maintain.There is a conflict in `spyder/plugins/ipythonconsole/widgets/client.py` compared to `4.x`. See [bnavigator:pr12534-close_leaks](https://github.com/bnavigator/spyder/tree/pr12534-close_leaks)

Observing some odd output:
```
...
[  161s] spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_stderr_file_remains_two_kernels PASSED [ 68%]
[  163s] spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_kernel_crash IPython: an enhanced interactive Python shell.
[  163s] 
[  163s] Subcommands
[  163s] -----------
[  163s] 
[  163s] Subcommands are launched as `ipython-kernel cmd [args]`. For information on
[  163s] using subcommand 'cmd', do: `ipython-kernel cmd -h`.
[  163s] 
[  163s] install
[  163s]     Install the IPython kernel
[  163s] 
[  163s] Options
[  163s] -------
[  163s] 
[  163s] Arguments that take values are actually convenience aliases to full
[  163s] Configurables, whose aliases are listed on the help line. For more information
[  163s] on full configurables, see '--help-all'.
[  163s] 
[  163s] --debug
[  163s]     set log level to logging.DEBUG (maximize logging output)
...
[  163s] 
[  163s] To see all available configurables, use `--help-all`
[  163s] 
[  250s] WARNING:traitlets:kernel died: 45.04048562049866
[  250s] WARNING:traitlets:kernel died: 45.04249310493469
[  250s] PASSED [ 68%]
[  250s] spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_remove_old_stderr_files SKIPPED [ 68%]
[  251s] spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_console_working_directory PASSED [ 68%]
...
```
Why is this "usage help text" printed to the shell immediately although pytest should capture all stdin and stderr? Also note the timeout and "kernel died" messages until the testing continues.


Unfortunately, the test_mainwindow suite still gets aborted:
```
[  588s] + python3 runtests.py --run-slow -k test_mainwindow
...
[ 1132s] spyder/app/tests/test_mainwindow.py::test_varexp_remove PASSED           [ 85%]
[ 1138s] spyder/app/tests/test_mainwindow.py::test_varexp_refresh /var/tmp/rpm-tmp.rc3qVM: line 78: 19514 Trace/breakpoint trap   (core dumped) python3 runtests.py --run-slow -k test_mainwindow
[ 1138s] error: Bad exit status from /var/tmp/rpm-tmp.rc3qVM (%check)

```


All other fast and slow tests which are not already excluded for a different reason (see comments in the `%check` section of the [specfile](https://build.opensuse.org/package/view_file/home:bnavigator:branches:devel:languages:python:numeric/spyder/spyder.spec?expand=1)) pass :tada:. Without this PR, some of the ipythonconsole tests would not.

*All of the above is about building locally. On [OBS](https://build.opensuse.org/package/show/home:bnavigator:branches:devel:languages:python:numeric/spyder), there is a new segfault which I have not investigated yet.*

**Comment by impact27** [comment]
> Could you please change the target branch from master to 4.x

Thanks @ccordoba12 

> Why is this "usage help text" printed to the shell immediately although pytest should capture all stdin and stderr? Also note the timeout and "kernel died" messages until the testing continues.

I think this text comes from another thread. "kernel died" for example comes from the heartbeat thread. If the kernel is closed and the heartbeat thread is somehow not closed, you would get these messages.

>Unfortunately, the test_mainwindow suite still gets aborted:

We are reusing the same window for the tests, but you could patch that to close the window and reopen it every time. Just dedent these two lines:

https://github.com/spyder-ide/spyder/blob/ae6f2947687af024377a2038fb88954eff6c75a8/spyder/app/tests/test_mainwindow.py#L223
https://github.com/spyder-ide/spyder/blob/ae6f2947687af024377a2038fb88954eff6c75a8/spyder/app/tests/test_mainwindow.py#L224

**Comment by bnavigator** [comment]
> We are reusing the same window for the tests, but you could patch that to close the window and reopen it every time. Just dedent these two lines:

Did not help. Just takes way longer (naturally).

Situation right now: If I run `test_runcell*` and `test_varexp*` in a separate pytest run, I only get the occasional abort, not always.

Your efforts tracking down the leaks definitely helped I lot. Thanks for the hard work @impact27!

**Comment by impact27** [comment]
> Situation right now: If I run `test_runcell*` and `test_varexp*` in a separate pytest run, I only get the occasional abort, not always.

So only those tests fail occasionnaly?

**Comment by bnavigator** [comment]
No, I think they are just the final straw that break the camel's back. The cause must be on a previous run, but I had no success of tracking it down so far.

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 1]
```diff
@@ -70,6 +70,21 @@ def __init__(self, *args, **kwargs):
         self._tmp_reading = False
         self._pdb_frame_loc = (None, None)
 
+    def will_close(self, externally_managed):
+        """
+        Close communication channels with the kernel if shutdown was not
+        called. If the kernel is not externally managed, shutdown the kernel
+        as well.
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/debugging.py:77]: This docstring doesn't seem to describe correctly what's happening in the code present in this method. Please update it.

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 2]
```diff
@@ -109,6 +111,46 @@ def __init__(self, ipyclient, additional_options, interpreter_versions,
             self.spyder_kernel_comm.register_call_handler(
                 request_id, handlers[request_id])
 
+    def __del__(self):
+        """Avoid destroying thread"""
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:115]:
```suggestion
        """Avoid destroying shutdown_thread."""
```

File: spyder/plugins/ipythonconsole/plugin.py [hunk 3]
```diff
@@ -1012,7 +1015,7 @@ def close_client(self, index=None, client=None, force=False):
 
         # if there aren't related clients we can remove stderr_file
         related_clients = self.get_related_clients(client)
-        if len(related_clients) == 0 and osp.exists(client.stderr_file):
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/ipythonconsole/plugin.py:1015]: There is a try: except already in `client.remove_stderr_file()`. Furthermore, I am not sure if the file handle can be open if the file doesn't exist, but if this is the case, this is needed to close the file handle properly.

[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/plugin.py:1015]: Ok, no problem then :+1:

**Comment by bnavigator** [comment]
Hi! Quick report, in case you are interested:

When 4.1.3 came out, I rebased this onto v4.1.3 and tried to build the package, but it hang during the test_mainwindow tests and never returned. So I tried without the patch and [this is how the 4.1.3 package is currently built](https://build.opensuse.org/package/show/openSUSE:Factory/spyder). Tests are only run on x86_64. [32-bit i586 segfaults](https://build.opensuse.org/project/show/home:bnavigator:spyderdevel) too often, probably because of ipython or QT.

Now with your recent commits I tried again. Both of the following scenarios work for x86_64:
- [Direct checkout of this PR's head](https://build.opensuse.org/package/show/home:bnavigator:spyderdevel/spyder-git-12534)
- [Rebased PR onto v4.1.3](https://build.opensuse.org/package/show/home:bnavigator:spyderdevel/spyder-413-pr12534), with additional patch from #12746 

Thanks for your hard work, @impact27!

**Pull request merged**